app/lib/dttp/batch_request.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module Dttp
+  class BatchRequest
+    class Error < StandardError; end
+
+    attr_reader :batch_id, :change_set_id, :headers, :change_sets
+
+    def initialize(batch_id: SecureRandom.uuid, change_set_id: SecureRandom.uuid)
+      @batch_id = batch_id
+      @change_set_id = change_set_id
+      @change_sets = []
+      @headers = { "Content-Type" => "multipart/mixed;boundary=batch_#{batch_id}" }
+    end
+
+    def add_change_set(entity:, payload:, content_id: SecureRandom.uuid)
+      @change_sets << { entity: entity, payload: payload, content_id: content_id }
+      content_id
+    end
+
+    def submit
+      response = Client.post("/$batch", body: body, headers: headers)
+      raise Error, response.body.gsub("\r", "") if response.code != 200
```

Comment:
I would of assumed success is 200 to 399

Reply:
The documentation and manual testing suggests `200` is what we should be looking out for and nothing else. The payload however gives move detail on the HTTP response codes for each change set request e.g.

```
--batchresponse_be1335fd-d817-4407-8724-4e5aa6ac88a1
Content-Type: multipart/mixed; boundary=changesetresponse_2c01fd6c-823c-418c-9aed-6aba0db0dcbd

--changesetresponse_2c01fd6c-823c-418c-9aed-6aba0db0dcbd
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 06910a13-3875-4d7b-9dbf-d3d651a18e45

HTTP/1.1 204 No Content
OData-Version: 4.0
Location: https://dttp-dev.api.crm4.dynamics.com/api/data/v9.1/contacts(243fc7f1-6035-eb11-a813-000d3ada6b1f)
OData-EntityId: https://dttp-dev.api.crm4.dynamics.com/api/data/v9.1/contacts(243fc7f1-6035-eb11-a813-000d3ada6b1f)


--changesetresponse_2c01fd6c-823c-418c-9aed-6aba0db0dcbd
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 939906f2-bbe3-4080-9ed0-1aa6660dfabb

HTTP/1.1 204 No Content
OData-Version: 4.0
Location: https://dttp-dev.api.crm4.dynamics.com/api/data/v9.1/dfe_placementassignments(273fc7f1-6035-eb11-a813-000d3ada6b1f)
OData-EntityId: https://dttp-dev.api.crm4.dynamics.com/api/data/v9.1/dfe_placementassignments(273fc7f1-6035-eb11-a813-000d3ada6b1f)


--changesetresponse_2c01fd6c-823c-418c-9aed-6aba0db0dcbd--
--batchresponse_be1335fd-d817-4407-8724-4e5aa6ac88a1--
```

Reply:
Umm does that mean the 200 is used for accepting the request, and the underlying can be failures?

http://docs.oasis-open.org/odata/odata/v4.0/os/part1-protocol/odata-v4.0-os-part1-protocol.html#_Toc372793753

Reply:
You would think so, but that's not the case. If there are errors in the underlying requests, the batch response code will be `400`. e.g.

```
--batchresponse_ee9fb140-0e8a-4bba-bc5e-d5040d28d538
Content-Type: multipart/mixed; boundary=changesetresponse_3877ce82-9f1f-4506-8aa3-0366e1b2f22c

--changesetresponse_3877ce82-9f1f-4506-8aa3-0366e1b2f22c
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: c218dfe2-0475-4cd4-bee1-618597c52495

HTTP/1.1 400 Bad Request
REQ_ID: 9dcaa8df-8b82-438a-a157-f0b7e7f6da63
Content-Type: application/json; odata.metadata=minimal
OData-Version: 4.0

{"error":{"code":"0x0","message":"The supplied reference link -- /dfe_ethnicities() -- is invalid. Expecting a reference link of the form /entityset(key)."}}
--changesetresponse_3877ce82-9f1f-4506-8aa3-0366e1b2f22c--
--batchresponse_ee9fb140-0e8a-4bba-bc5e-d5040d28d538--
```